### PR TITLE
Supporting ESM files with `.js` extension

### LINKED ISSUE #29: Support .js config files in "type": "module" projects?
Hi there,

would you consider supporting `.js` files inside projects with `"type": "module"` set? Otherwise is pretty hard to provide config files sometimes, because not every user of `lilconfig` actually includes `.cjs` as an allowed file extension.

Currently I'm using the following patch, which should work for CommonJS and ECMAScript Modules since [Node 12](https://nodejs.org/dist/latest-v12.x/docs/api/esm.html#esm_import_expressions), so it would require a semver-major bump (or a fallback for Node 10):

```diff
diff --git a/dist/index.js b/dist/index.js
index a153607e305eb6d40b735aa176eee388b8172dc9..497efc7dc60165b306f46211fb7ddd811b51ef72 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -28,7 +28,7 @@ function getSearchPaths(startDir, stopDir) {
     }, { searchPlaces: [], passedStopDir: false }).searchPlaces;
 }
 exports.defaultLoaders = Object.freeze({
-    '.js': require,
+    '.js': id => import(id).then(mod => mod.default),
     '.json': require,
     '.cjs': require,
     noExt(_, content) {
@@ -123,7 +123,7 @@ function lilconfig(name, options) {
                 }
                 else {
                     validateLoader(loader, loaderKey);
-                    result.config = loader(filepath, content);
+                    result.config = await loader(filepath, content);
                 }
                 result.filepath = filepath;
                 break;
```

### PR DESCRIPTION
**Pull request by marekdedic**
Closes #29 

I have verified that this also works with CommonJS in `.js` files - it's decided by the `type` property in `package.json`.

### CODE REVIEW

**Comment by marekdedic** [comment]
Btw, `getDefaultSearchPlaces` could also use an update - see https://github.com/cosmiconfig/cosmiconfig/blob/a5a842547c13392ebb89a485b9e56d9f37e3cbd3/src/defaults.ts#L12-L32

**Comment by antonk52** [comment]
Thanks for the PR.

it makes sense to update the default search places in the same PR. Specifically adding support for .mjs extension. Keeping in mind that it would not work in the sync version

**Comment by marekdedic** [comment]
Done :)

File: src/index.ts [hunk 1]
```diff
@@ -45,15 +45,26 @@ export interface OptionsSync extends OptionsBase {
...
+        `.config/${name}rc.yaml`,
+        `.config/${name}rc.yml`,
         `.config/${name}rc.js`,
+        `.config/${name}rc.ts`,
         `.config/${name}rc.cjs`,
+        `.config/${name}rc.mjs`,
         `${name}.config.js`,
+        `${name}.config.ts`,
         `${name}.config.cjs`,
+        `${name}.config.mjs`,
```
Comment (antonk52) [inline src/index.ts:67]: This is tricky. As part of this PR it makes sense to only add `.mjs` extension. 

Yaml and ts extensions are not and won't be supported out of the box so they should all be removed. Let's also remove `.${name}rc` as it is an unrelated change. Sorry for the confusion as I meant that only `.mjs` extension should be added in this PR. This PR would also require to add a couple of test cases. One for an `.mjs` file and one for js file with `esm default export`.

We should also consider how cosmiconfig handles esm files when sync api is used. they have a [section](https://www.npmjs.com/package/cosmiconfig#loading-js-modules) about this in the docs. I haven't checked the implementation yet. Can do that in the next day or two.

Comment (marekdedic) [reply src/index.ts]: Hi,
thanks, I tried to add some `mjs` tests, but they don't work and I'm not sure what the correct behaviour even is.

Could you please review and advise?

**Comment by antonk52** [comment]
Apologies for the delay. I still intend to merge this PR and release the changes. I had a medical emergency earlier and have trouble with cognitive tasks currently. I will try to test these changes some time later next week. Specifically there needs to be some testing and test cases done on using lilconfig searching for config files in projects with package.json#type:module and old CJS projects. We also need to compare the output of searching for .mjs files with sync searcher and comparing the behaviour with cosmiconfig. Once we are sure that the behaviour is fairly similar, I think it will be safe to merge this. 

Thank you for your patience.

**Comment by marekdedic** [comment]
Hi,
no worries, health comes first!

**Comment by antonk52** [comment]
Hi @marekdedic, thank you for your patience. I finally got a chance to look into this feature and it turns out more complex than the initial impression. I have opened another PR #46 with all the documented changes that I will continue to work until it feels like it is ready to be merged. I am going to close this PR. Having said that I do not want you to think that your time was wasted. Thanks to your PR this feature will be soon added to lilconfig and I would like to add your credentials in the commit message and release notes if that is okay with you, let me know. I would appreciate your feedback on #46.

Many thanks for taking your time and pushing this forward!

**Comment by marekdedic** [comment]
Hi, thanks for the heads up and further work on this :) I'll be looking forward to #46

**Pull request closed without merging**